tx_test.go

```diff
@@ -18,6 +21,107 @@ import (
 	"go.etcd.io/bbolt/internal/btesting"
 )
 
+func TestTx_Check_Page(t *testing.T) {
+	bucketKey := "testBucket"
+	db := btesting.MustCreateDBWithOption(t, &bolt.Options{PageSize: pageSize})
+	defer db.MustDeleteFile()
+
+	// arrange
+	if err := db.Update(func(tx *bolt.Tx) error {
+		b, err := tx.CreateBucketIfNotExists([]byte(bucketKey))
+		if err != nil {
+			t.Fatal(err)
+		}
+		// insert random data
+		for i := 0; i < 10; i++ {
+			generateSampleDataInBucket(t, b, 3)
+		}
+		return nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+	db.MustCheck()
+
+	// act : corrupt a leaf page
+	corruptLeafPageFromDB(t, db.Path())
+
+	// assert: expected an error after check
+	if err := db.View(func(tx *bolt.Tx) error {
+		for err := range tx.Check() {
+			fmt.Println(err)
+		}
+		return nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+	db.MustClose()
+}
+
+func corruptLeafPageFromDB(t testing.TB, dbPath string) {
+	_, hwm, err := guts_cli.ReadPageAndHWMSize(dbPath)
+	if err != nil {
+		t.Fatalf("error reading HWM from dbPath %v: %v", dbPath, err)
+	}
+
+	var victimPage *common.Page
+	var victimBuf []byte
+
+	// find first leaf page starting page after root bucket's page
+	for pageID := uint64(4); pageID < uint64(hwm); {
+		p, buf, err := guts_cli.ReadPage(dbPath, pageID)
+		if err != nil {
+			t.Fatalf("error reading page %v from dbPath %v: %v", pageID, dbPath, err)
+		}
+
+		if p.IsLeafPage() {
+			victimPage = p
+			victimBuf = buf
+			break
+		}
+	}
+
+	if victimPage == nil || victimBuf == nil {
+		t.Fatalf("no leaf page is found in dbPath %v", dbPath)
+	}
+
+	// corrupt page's type
+	invalidLPE := common.NewLeafPageElement(0, 0, 0, 0)
```

Comment:
@fuweid i used this as invalid `LeafPageElement` but still panics :/ 

```
=== RUN   TestTx_Check_Page
    btesting.go:47: Opening bbolt DB at: /var/folders/vh/wkbmyt411316pt46h81d6_300000gn/T/TestTx_Check_Page2552757781/001/db
panic: runtime error: index out of range [0] with length 0
```

Reply:
You should start with index 16 instead of 17.
The panic is expected because bbolt doesn't verify these information before read.

And you should use memory byte order instead of using `Big-endian`.

Reply:
done